src/ManifoldsBase.jl

```diff
@@ -425,6 +425,35 @@ See also: [`EmbeddedManifold`](@ref), [`project!`](@ref project!(M::AbstractMani
 """
 embed!(M::AbstractManifold, Y, p, X) = copyto!(M, Y, p, X)
 
+"""
+    embed_project(M::AbstractManifold, p)
+
+Embed `p` from manifold `M` an project it back to `M`. For points from `M` this is identity
+but in case embedding is defined for points outside of `M`, this can serve as a way
+to for example remove numerical innacuracies caused by some algorithms.
+"""
+function embed_project(M::AbstractManifold, p)
+    return project(M, embed(M, p))
+end
+"""
+    embed_project(M::AbstractManifold, p, X)
+
+Embed vector `X` tangent at `p` from manifold `M` an project it back to tangent space
+at `p`. For points from that tangent space this is identity but in case embedding is
+defined for tagent vectors from outside of it, this can serve as a way to for example remove
+numerical innacuracies caused by some algorithms.
+"""
+function embed_project(M::AbstractManifold, p, X)
```

Comment:
This should fix https://github.com/JuliaManifolds/Manifolds.jl/issues/522 . I've added it here because I'm definitely expecting this PR to be breaking anyway so it's going to be large-ish anyway.

Reply:
I'm also going to use this as a ManifoldsBase.jl replacement for `retract!` and `project_tangent!` in Optim.jl: https://github.com/JuliaNLSolvers/Optim.jl/blob/master/src/Manifolds.jl .

Reply:
I am not sure why we need this new function, but ok;

With the `retract!` from Optim.jl be careful that the retract there is not our retract (but basically “just” a projection) if I remember the discussion correctly from a longer discussion with Antoine (since he used the topological retraction in there if I remember correctly)

Reply:
For the breaking-ness – I had hoped this would be possible in a non-breaking way, but sure if it breaks, then it breaks. I would then just like to be a bit more precise in the release what actually brakes and how to fix things (we did not do that much before, but I feel it might be nice also for us to remember).

Reply:
> I am not sure why we need this new function, but ok;
> 
> With the `retract!` from Optim.jl be careful that the retract there is not our retract (but basically “just” a projection) if I remember the discussion correctly from a longer discussion with Antoine (since he used the topological retraction in there if I remember correctly)

If I understand right, Optim's `retract is `embed_project` for points and `project_tangent` is `embed_project` for tangent vectors -- if you think I'm wrong, then I'd like to know :slightly_smiling_face: .

> For the breaking-ness – I had hoped this would be possible in a non-breaking way, but sure if it breaks, then it breaks. I would then just like to be a bit more precise in the release what actually brakes and how to fix things (we did not do that much before, but I feel it might be nice also for us to remember).

Sure, I will make a list of changes when I get Manifolds.jl running.